File: JavaScript/sort/StoogeSort.js
/*

Stooge sort is a recursive sorting algorithm.
It is notable for its exceptionally bad time complexity which is slower compared to reasonable sorting algorithms.

Algorithm - 1) If value of the start is larger than the value at the end, swap them
            2) If there are 3 or more elements in the list, then -
                a] Stooge sort the initial 2/3 of the list
                b] Stooge sort the final 2/3 of the list
                c] Stooge sort the initial 2/3 of the list again

Time Complexity - O(n(log3/log1.5)) ( Slower than bubble sort )

*/

// Prompt node package for taking user input
const prompt = require("prompt-sync")({ sigint: true });

function stoogeSort(array, start, end) {
  let pivot;

  // Checking condition - whehther there are at least 2 elements or not
  if (end - start + 1 > 2) {
    pivot = Math.floor((end - start + 1) / 3);
    stoogeSort(array, start, end - pivot); // sort the first 2/3
    stoogeSort(array, start + pivot, end); // sort the last 2/3
    stoogeSort(array, start, end - pivot); // sort the first 2/3 again
  }

  // Swap if the leftmost element is larget than rightmost
  if (array[start] > array[end]) {
    [array[start], array[end]] = [array[end], array[start]];
  }
}

/* Workflow of user input */

// Take array length as input
let arrayLength = +prompt("Enter array length - ");

// Check whether the entered value is number or not
if (isNaN(arrayLength)) return console.log("Only numbers are allowed");

// Globally declared array
let array = [];

// Take array items
for (let i = 1; i <= arrayLength; i++) {
  array.push(+prompt(`Enter ${i} element - `));

  // Check whether the entered value is number or not
  if (array.includes(NaN)) return console.log("Only numbers are allowed");
}

console.log("Your array - ", array);

// Call the algorithm
stoogeSort(array, 0, arrayLength - 1);
console.log("Sorted Array - ", array);

// Sample I/O

/*

> node StoogeSort

Enter array length - 5
Enter 1 element - 45
Enter 2 element - 86
Enter 3 element - 23
Enter 4 element - 98
Enter 5 element - 66
Your array -  [ 45, 86, 23, 98, 66 ]
Sorted Array -  [ 23, 45, 66, 86, 98 ]

*/

File: C/sort/stooge_sort.c
// C program to implement Stooge  Sort
/*
The function of the stooge sort function is to check if the value at index 0 is 
greater than the value at last index, if YES then to swap those values.

Call the Stooge sort function recursively on initial 2/3rd of the array, last 2/3rd
and again initial 2/3rd to get the given array sorted
*/
#include <stdio.h>

// Helper function to swap two values
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

void stooge_sort(int arr[], int start, int end)
{
    // If the value at last index is smaller than the value at index 0, Swap them.
    if (arr[start] > arr[end])
        swap(&arr[start], &arr[end]);

    // For finding the first and last two by third elements in the array
    if (end - start + 1 > 2)
    {
        int twobythird = (end - start + 1) / 3;

        //Recursively call the function on the initial two by third, last two by third followed by the initial two by third
        if ((end - twobythird) >= start)
            stooge_sort(arr, start, end - twobythird);
        if (end >= (start + twobythird))
            stooge_sort(arr, start + twobythird, end);
        if ((end - twobythird) >= start)
            stooge_sort(arr, start, end - twobythird);
    }
}

int main()
{
    int n;
    printf("\nHow many numbers do you want to sort? ");
    scanf("%d", &n);
    int arr[n];

    if (n <= 0)
    {
        printf("There are no numbers to sort!!!");
        return 0;
    }
    // Input the numbers to sort
    printf("Enter the numbers: ");
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    //Call the sort function
    stooge_sort(arr, 0, n - 1);

    printf("The numbers in sorted order is: ");
    // Print the sorted array
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");

    return 0;
}

/*
Time Complexity: O(n^2.7095), slower than bubble sort
Space Complexity: O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE 1

How many numbers do you want to sort? 5
Enter the numbers: 1 3 5 2 4
The numbers in sorted order is:  1  2  3  4  5

SAMPLE 2

How many numbers do you want to sort? 0
There are no numbers to sort!!!

*/
